proxy/src/redis/kv_ops.rs

```diff
@@ -0,0 +1,262 @@
+use redis::{AsyncCommands, ToRedisArgs};
+use std::convert::Infallible;
+use tokio::sync::{mpsc, oneshot};
+
+use super::connection_with_credentials_provider::ConnectionWithCredentialsProvider;
+use crate::metrics::CancelChannelSizeGuard;
+use crate::rate_limiter::{GlobalRateLimiter, RateBucketInfo};
+
+// Message types for sending through mpsc channel
+pub enum RedisOp {
+    HSet {
+        key: String,
+        field: String,
+        value: String,
+        resp_tx: Option<oneshot::Sender<anyhow::Result<()>>>,
+        _guard: CancelChannelSizeGuard<'static>,
+    },
+    HSetMultiple {
+        key: String,
+        items: Vec<(String, String)>,
+        resp_tx: Option<oneshot::Sender<anyhow::Result<()>>>,
+        _guard: CancelChannelSizeGuard<'static>,
+    },
+    HGet {
+        key: String,
+        field: String,
+        resp_tx: oneshot::Sender<anyhow::Result<String>>,
+        _guard: CancelChannelSizeGuard<'static>,
+    },
+    HGetAll {
+        key: String,
+        resp_tx: oneshot::Sender<anyhow::Result<Vec<(String, String)>>>,
+        _guard: CancelChannelSizeGuard<'static>,
+    },
+    HDel {
+        key: String,
+        field: String,
+        resp_tx: Option<oneshot::Sender<anyhow::Result<()>>>,
+        _guard: CancelChannelSizeGuard<'static>,
+    },
+}
+
+pub struct RedisKVClient {
+    client: ConnectionWithCredentialsProvider,
+    limiter: GlobalRateLimiter,
+    pub rx: mpsc::Receiver<RedisOp>, // recieve messages from other tasks
+}
+
+impl RedisKVClient {
+    pub fn new(
+        client: ConnectionWithCredentialsProvider,
+        info: &'static [RateBucketInfo],
+        rx: mpsc::Receiver<RedisOp>,
+    ) -> anyhow::Result<Self> {
+        Ok(Self {
+            client,
+            limiter: GlobalRateLimiter::new(info.into()),
+            rx,
+        })
+    }
+
+    pub async fn try_connect(&mut self) -> anyhow::Result<()> {
+        match self.client.connect().await {
+            Ok(()) => {}
+            Err(e) => {
+                tracing::error!("failed to connect to redis: {e}");
+                return Err(e);
+            }
+        }
+        Ok(())
+    }
+
+    // Running as a separate task to accept messages through the rx channel
+    pub async fn handle_messages(&mut self) -> anyhow::Result<Infallible> {
+        loop {
+            if let Some(msg) = self.rx.recv().await {
```

Comment:
More of a comment for the future.

If we do observe throughput issues, we can try make use of https://docs.rs/tokio/latest/tokio/sync/mpsc/struct.Receiver.html#method.recv_many + https://docs.rs/redis/latest/redis/#pipelining to try and amortise the round-trip time.

Reply:
added comment there